### PR TITLE
[RNMobile] Embed block: Unavailable preview fallback

### PR DESCRIPTION
**Pull request by fluiddot**
⚠️ **Important note:** [✅  Reverted in [this commit](https://github.com/WordPress/gutenberg/pull/33266/commits/b89aaab11e94798f1edd678a1c53f7366c270c1f)] Please re-enable the `dev` flag for the embed block before merging this PR (https://github.com/WordPress/gutenberg/pull/33266#discussion_r665999208).

`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3710
`WordPress-Android` PR: https://github.com/wordpress-mobile/WordPress-Android/pull/15001
`WordPress-iOS` PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/16825

## Description

This PR adds the unavailable preview fallback component to the embed block. Additionally, I made a small refactor on the styles of this block to match the original designs.

### Post
1. Add an Embed block to a post
2. Set a valid embed URL without thumbnail (i.e.: https://twitter.com/wordpressdotcom/status/1388164029840044042)
3. Observe that the block's content displays a message that inline preview is coming soon and the text "PREVIEW POST"
4. Tap on the block to edit it
5. Observe that an info modal shows up explaining that previews are coming soon
6. Observe that tapping on the dismiss button closes the modal
7. Tap on the block again to open the previous modal
8. Tap on "Preview post"
9. Observe that the modal is closed and the preview modal is opened displaying the embed block

### Page
1. Add an Embed block to a page
2. Set a valid embed URL without thumbnail (i.e.: https://twitter.com/wordpressdotcom/status/1388164029840044042)
3. Observe that the block's content displays a message that inline preview is coming soon and the text "PREVIEW PAGE"
4. Tap on the block to edit it
5. Observe that an info modal shows up explaining that previews are coming soon
6. Observe that tapping on the dismiss button closes the modal
7. Tap on the block again to open the previous modal
8. Tap on "Preview page"
9. Observe that the modal is closed and the preview modal is opened displaying the embed block

### iOS
Light mode|Light mode|Dark mode|Dark mode
--|--|--|--
|||

### Android
Light mode|Light mode|Dark mode|Dark mode
--|--|--|--
|||

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/index.native.js [hunk 1]
```diff
@@ -266,7 +266,7 @@ export const registerCoreBlocks = () => {
 		audio,
 		reusableBlock,
 		search,
-		devOnly( embed ),
+		embed,
```
Comment (fluiddot) [inline packages/block-library/src/index.native.js]: **⚠️ I enabled the block just for testing via the installable builds, once we finish the testing we should revert this change.**

**Comment by hypest** [comment]
👋 @fluiddot ! Gave the preview fallback a spin on WPAndroid and here's a couple of things I noticed:

1. There's seems to still be a shade overlaid after finishing inputting the URL, see the video below. The UI is actually tappable but the shade is there.

https://user-images.githubusercontent.com/1032913/124923041-c08db180-e002-11eb-84b1-910e425bec04.mp4


2. I see a UI glitch happening where the UI is not properly updated after going to the preview screen and a rotation has happened before returning, see the video below. I don't think it's an issue introduced with this set of PR but since we're going to make the Preview functionality even more prominent now, we should try and improve the update logic.


https://user-images.githubusercontent.com/1032913/124923857-84a71c00-e003-11eb-93d2-2f7d5435a05e.mov

3. Using the replace URL functionality, I see that the block state seems to change to the one when the block is empty, see video below. Is that expected? I feel that the block would stay as is until I commit the URL change 🤔


https://user-images.githubusercontent.com/1032913/124925440-1bc0a380-e005-11eb-94c8-918175a36d16.mp4

**Comment by ceyhun** [comment]
Thanks @fluiddot! I've tested on iOS and it's working great. I couldn't reproduce the shade (issue 1 described above) or the glitch (issue 2) on iOS, though the state change (3) is there.

One little thing that caught my eye, I think there was a question mark icon at the top right of the block in the designs.

embed-preview-question-mark

**Comment by fluiddot** [comment]
> 👋 @fluiddot ! Gave the preview fallback a spin on WPAndroid and here's a couple of things I noticed:
> 
> 1. There's seems to still be a shade overlaid after finishing inputting the URL, see the video below. The UI is actually tappable but the shade is there.

I can't manage to reproduce it but I'll try to debug it, did it happen every time you set an URL?

> 2. I see a UI glitch happening where the UI is not properly updated after going to the preview screen and a rotation has happened before returning, see the video below. I don't think it's an issue introduced with this set of PR but since we're going to make the Preview functionality even more prominent now, we should try and improve the update logic.

Oh, I didn't notice that, I'm trying to reproduce it but no luck so far, could you share the steps you followed?

> 3. Using the replace URL functionality, I see that the block state seems to change to the one when the block is empty, see video below. Is that expected? I feel that the block would stay as is until I commit the URL change 🤔

Right, this is caused by how the flow works on the web version, when editing it's rendering the embed placeholder which doesn't include the preview. If we want to keep the preview while editing, we should refactor the edit component and most likely create a new one for this new flow.

**Replace flow in the web version:**

https://user-images.githubusercontent.com/14905380/124935929-62fb6400-e006-11eb-9902-3dd075022c1b.mp4

**Comment by fluiddot** [comment]
> Thanks @fluiddot! I've tested on iOS and it's working great. I couldn't reproduce the shade (issue 1 described above) or the glitch (issue 2) on iOS, though the state change (3) is there.
> 
> One little thing that caught my eye, I think there was a question mark icon at the top right of the block in the designs.
> 
> embed-preview-question-mark

Good catch! I overlooked that element, I'll update the block to include it, thanks for spotting it 🙇 !

**Comment by hypest** [comment]
> Replace flow in the web version:

Aha, yeah on the web the editing happens inline and replaces the preview. On native mobile, since the input happens out-of-bound in a bottomsheet, we need to refactor the implementation and avoid replacing the preview.

**Comment by fluiddot** [comment]
> > Replace flow in the web version:
> 
> Aha, yeah on the web the editing happens inline and replaces the preview. On native mobile, since the input happens out-of-bound in a bottomsheet, we need to refactor the implementation and avoid replacing the preview.

Ok, I'll give it a spin and figure out a way to include it 👍 .

**Comment by hypest** [comment]
> I can't manage to reproduce it but I'll try to debug it, did it happen every time you set an URL?

No, it doesn't happen consistently.

> Oh, I didn't notice that, I'm trying to reproduce it but no luck so far, could you share the steps you followed?

Looks like I **can't replicate** on the [installable build](https://github.com/wordpress-mobile/WordPress-Android/pull/15001#issuecomment-876264896) but I can on the locally built wasabiDebug build, here are the steps:

1. In a new post, add two embed blocks, the first one pointing to [this YT video](https://www.youtube.com/watch?v=UZ4yaV9uepA) and the second one pointing to the thumbnail-less URL shared in the PR instructions (https://twitter.com/wordpressdotcom/status/1388164029840044042).
2. Turn the device to landscape mode.
3. Tap on the "PREVIEW POST" link-button in the second embed block to launch the preview.
4. While still in the preview post screen, turn the device back to portrait.
5. Exit the preview screen to return to the editor.
6. ⚠️Notice how the YT preview is overflowing the device screen horizontally. Sometimes, the second block is overlapping the first one too.

I understand that this could also be related to device screen metrics actually. I'm testing on a Pixel 2XL, and it looks like the YT preview component doesn't adjust back to smaller dimensions after returning from the  preview screen.

File: packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java [hunk 2]
```diff
@@ -240,6 +240,11 @@ public void gutenbergDidSendButtonPressedAction(String buttonType) {
 
             }
 
+            @Override
+            public void requestPreview() {
+
```
Comment (hypest) [inline packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java]: Want to do something like https://github.com/WordPress/gutenberg/blob/27fc8980ddf875757e671688f83a9bfbfa73a0c5/packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java#L235 and put out a Toast popup message here?

**Comment by fluiddot** [comment]
> > Thanks @fluiddot! I've tested on iOS and it's working great. I couldn't reproduce the shade (issue 1 described above) or the glitch (issue 2) on iOS, though the state change (3) is there.
> > One little thing that caught my eye, I think there was a question mark icon at the top right of the block in the designs.
> > embed-preview-question-mark
> 
> Good catch! I overlooked that element, I'll update the block to include it, thanks for spotting it 🙇 !

I've just included the help icon in these commits ([#1](https://github.com/WordPress/gutenberg/pull/33266/commits/5c1282a090d823f9fd49b621fd6fc715bd5c5933) and [#2](https://github.com/WordPress/gutenberg/pull/33266/commits/27fc8980ddf875757e671688f83a9bfbfa73a0c5)), here you can see how it looks like:

<details><summary><strong>Screenshots</strong></summary>

### iOS
Light mode|Dark mode
--|--
|

### Android
Light mode|Dark mode
--|--
|

</details>

File: packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java (hunk 2 above)
Comment (fluiddot) [reply packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java]: Yeah, I like the idea of notifying somehow that a mocked native call has been called, it could be useful when testing on the demo project 👍 .

**Comment by hypest** [comment]
Did a first pass to the Android side of the bridge changes here @fluiddot and looks good to me 👍. Happy to make another pass when the PR stabilizes even more (I'm referring to https://github.com/WordPress/gutenberg/pull/33266#issuecomment-876491927 and perhaps there might be more bridge changes).

**Comment by fluiddot** [comment]
> > I can't manage to reproduce it but I'll try to debug it, did it happen every time you set an URL?
> 
> No, it doesn't happen consistently.

I'm wondering if this issue can be reproduced in other places where we use the Link settings or is only related to the embed block 🤔 .

> > Oh, I didn't notice that, I'm trying to reproduce it but no luck so far, could you share the steps you followed?
> 
> Looks like I **can't replicate** on the [installable build](https://github.com/wordpress-mobile/WordPress-Android/pull/15001#issuecomment-876264896) but I can on the locally built wasabiDebug build, here are the steps:
> 
> 1. In a new post, add two embed blocks, the first one pointing to [this YT video](https://www.youtube.com/watch?v=UZ4yaV9uepA) and the second one pointing to the thumbnail-less URL shared in the PR instructions (https://twitter.com/wordpressdotcom/status/1388164029840044042).
> 2. Turn the device to landscape mode.
> 3. Tap on the "PREVIEW POST" link-button in the second embed block to launch the preview.
> 4. While still in the preview post screen, turn the device back to portrait.
> 5. Exit the preview screen to return to the editor.
> 6. ⚠️Notice how the YT preview is overflowing the device screen horizontally. Sometimes, the second block is overlapping the first one too.
> 
> I understand that this could also be related to device screen metrics actually. I'm testing on a Pixel 2XL, and it looks like the YT preview component doesn't adjust back to smaller dimensions after returning from the preview screen.

Interesting, I only managed to reproduce it with a local build too, in fact, only rotating the device from landscape to portrait also causes the issue without having to preview the post.

**Comment by fluiddot** [comment]
I've done a refactor of the embed block and so far, looks like all issues are addressed:

## Backdrop of bottom sheet modal is displayed after dismissing it

**Description:** There's seems to still be a shade overlaid after finishing inputting the URL, see the video below. The UI is actually tappable but the shade is there.

**Solution:** I couldn't reproduce it and looks like it's not happening consistently. For now, I wouldn't prioritize it unless we figure out a reliable way to reproduce it and then debug it.

## UI glitch when rotating device

**Description:** I see a UI glitch happening where the UI is not properly updated after going to the preview screen and a rotation has happened before returning, see the video below. I don't think it's an issue introduced with this set of PR but since we're going to make the Preview functionality even more prominent now, we should try and improve the update logic.

**Solution:** I noticed that the styles applied in the image component used to render the thumbnail was causing the issue so I made a small tweak in the flex values.

**Fixed in:** https://github.com/WordPress/gutenberg/pull/33266/commits/b5c0bdbe60005dce137d57429d92a4474b97387d

## Block state changes when replacing the URL 

**Description:** Using the replace URL functionality, I see that the block state seems to change to the one when the block is empty, see video below. Is that expected? I feel that the block would stay as is until I commit the URL change

**Solution:** I created a native version of the edit component with a different behavior to render the placeholder and the preview components.

**Fixed in:** https://github.com/WordPress/gutenberg/pull/33266/commits/f12ad72bf1553322d1c3624372d017131434f601

File: packages/block-library/src/embed/edit.native.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+			// The external oEmbed provider does not exist. We got no type info and no html.
+			const badEmbedProvider =
+				embedPreview?.html === false &&
+				embedPreview?.type === undefined;
+			// Some WordPress URLs that can't be embedded will cause the API to return
+			// a valid JSON response with no HTML and `data.status` set to 404, rather
+			// than generating a fallback response as other embeds do.
+			const wordpressCantEmbed = Platform.select( {
+				web: embedPreview?.data?.status === 404,
```
Comment (ceyhun) [inline packages/block-library/src/embed/edit.native.js]: I think we don't need this anymore as well. We can maybe also remove the `native` part from the web version `edit.js`.

**Review by ceyhun — APPROVED** [review]
I confirmed that the 2. and 3. issues are resolved, but I encountered the 1. issue as well on WPAndroid, though I'm not sure how to reproduce it or if it's even related to the changes here. Things are working as expected on WPiOS. Just have a small comment, but other then that I think it LGTM!

Edit: Small heads up about reverting the change removing `devOnly` flag before merging.

File: packages/block-library/src/embed/edit.native.js (hunk 3 above)
Comment (fluiddot) [reply packages/block-library/src/embed/edit.native.js]: Good catch, I totally overlooked this part, I'll update both files 🙇 .

**Comment by fluiddot** [comment]
> I confirmed that the 2. and 3. issues are resolved, but I encountered the 1. issue as well on WPAndroid, though I'm not sure how to reproduce it or if it's even related to the changes here. Things are working as expected on WPiOS. Just have a small comment, but other then that I think it LGTM!

I couldn't reproduce point 1 yet but I'm wondering if, as you commented, it's actually related to these changes 🤔 . I'm thinking to open an issue as a follow-up and tackle it in a different PR.

> Edit: Small heads up about reverting the change removing `devOnly` flag before merging.

Thanks for the heads up! I'll revert it now that the PR is approved 👍 .

File: packages/block-library/src/embed/edit.native.js (hunk 3 above)
Comment (fluiddot) [reply packages/block-library/src/embed/edit.native.js]: I've just applied the update in [this commit](https://github.com/WordPress/gutenberg/pull/33266/commits/9a4031314fb13f518c40479d7783759e87540fb6).

File: packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java (hunk 2 above)
Comment (fluiddot) [reply packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java]: I applied the suggestion in [this commit](https://github.com/WordPress/gutenberg/pull/33266/commits/2bee10d420c893aa4f1d883be1143bfeccea222e).
